crud/src/main/java/com/redhat/lightblue/crud/mongo/Translator.java

```diff
@@ -380,21 +381,31 @@ public DBObject translate(EntityMetadata md, UpdateExpression expr)
                                               Sort s) {
         Set<Path> fields=new HashSet<>();
         FieldCursor cursor=md.getFieldCursor();
-        while(cursor.next()) {
-            Path field=cursor.getCurrentPath();
-            FieldTreeNode node=cursor.getCurrentNode();
-            if( (node instanceof ObjectField) ||
-                (node instanceof ArrayField &&
-                 ((ArrayField)node).getElement() instanceof ObjectArrayElement) ) {
-                // include its member fields
-            } else if( (p!=null && p.isFieldRequiredToEvaluateProjection(field)) ||
-                       (q!=null && q.isRequired(field)) ||
-                       (s!=null && s.isRequired(field)) ) {
-                LOGGER.debug("{}: required",field);
-                fields.add(field);
-            } else {
-                LOGGER.debug("{}: not required", field);
-            }
+        if(cursor.next()) {
+            boolean done=false;
+            do {
+                Path field=cursor.getCurrentPath();
+                FieldTreeNode node=cursor.getCurrentNode();
+                LOGGER.debug("Checking if {} is included ({})",field,node);
+                if(node instanceof ResolvedReferenceField||
+                   node instanceof ReferenceField) {
+                    done=!cursor.nextSibling();
```

Comment:
Couldn't this stop too early?  What if the ref field is in the middle of the cursor?  What if there are multiple references in the query?